src/mlang/m_ir/mir.ml

```diff
@@ -411,7 +411,53 @@ let find_var_by_name (p : program) (name : string Pos.marked) : Variable.t =
            (Pos.VarNameToID.find name p.program_idmap))
     with Not_found -> Errors.raise_spanned_error "unknown variable" (Pos.get_position name))
 
-let find_vars_by_io (p : program) (io_to_find : io) : Variable.t list =
-  VariableMap.fold
-    (fun k v acc -> if v.var_io = io_to_find then k :: acc else acc)
-    p.program_vars []
+let find_vars_by_io (p : program) (io_to_find : io) : rule_data RuleMap.t =
+  let is_var_io (_, var_data) = var_data.var_io = io_to_find in
+  RuleMap.filter_map
+    (fun _ y ->
+      let rule_vars = List.filter is_var_io y.rule_vars in
+      Some { y with rule_vars })
+    p.program_rules
```

Comment:
I would expect this function to return a `VarDict.t`, as the name is similar to the other helper function of this kind.
If you actually need other information of `var_definition`, I think the name would need to change.

(But yes, dissociating `var_definition` from the actual variable map is cumbersome as we see in the next function below.)